contracts/AccessControl/RoleBasedAccessControl.sol

```diff
@@ -0,0 +1,147 @@
+    // Types
+    ////////////////
+
+    // Łukasiewicz logic values
+    enum TriState {
+        Unset,
+        True,
+        False
+    }
+
+    ////////////////
+    // Constants
+    ////////////////
+
+    IAccessControlled public constant GLOBAL = IAccessControlled(0x0);
+
+    ////////////////
+    // State
+    ////////////////
+
+    mapping (address => mapping(bytes32 => mapping(address => TriState))) access;
+
+    // TODO: Retrieve list of contracts
+    // TODO: Retrieve list of users per role
+
+    ////////////////
+    // Events
+    ////////////////
+
+    event AccessChanged(
+        address controler,
+        address subject,
+        bytes32 role,
+        IAccessControlled object,
+        TriState oldValue,
+        TriState newValue
+    );
+
+    event Access(
+        address subject,
+        bytes32 role,
+        IAccessControlled object,
+        bytes4 verb,
+        bool granted
+    );
+
+    ////////////////
+    // Constructor
+    ////////////////
+
+    function RoleBasedAccessControl()
+        AccessControlled(this) // We are our own policy. This is immutable.
+    {
+        // Issue the glboal AccessContoler role to creator
+        access[msg.sender][ROLE_ACCESS_CONTROLER][GLOBAL] = TriState.True;
+    }
+
+    ////////////////
+    // Public functions
+    ////////////////
+
+    // Overrides `AccessControlled.setAccessPolicy(IAccessPolicy)`
+    function setAccessPolicy(IAccessPolicy)
+        public
+        only(ROLE_ACCESS_CONTROLER)
+    {
+        // `RoleBasedAccessControl` always controls its
+        // own access. Disallow changing this by overriding
+        // the `AccessControlled.setAccessPolicy` function.
+        revert();
+    }
+
+    // Implements `IAccessPolicy.allowed(IAccessControlled, bytes32, address, bytes4)`
+    function allowed(
+        address subject,
+        bytes32 role,
+        IAccessControlled object,
+        bytes4 verb
+    )
+        public
+        // constant // NOTE: Solidity does not allow subtyping interfaces
+        returns (bool)
+    {
+        // Try local access first
+        TriState localAccess = access[subject][role][object];
+        if (localAccess != TriState.Unset) {
+            bool grantedLocal = localAccess == TriState.True;
+
+            // Log and return
+            Access(subject, role, object, verb, grantedLocal);
+            return grantedLocal;
+        }
+
+        // Try global state
+        TriState globalAccess = access[subject][role][GLOBAL];
+        bool granted = globalAccess == TriState.True;
+
+        // Log and return
+        Access(subject, role, object, verb, granted);
+        return granted;
+    }
+
+    // Assign a role to a user globally
+    function setUserRole(
+        address subject,
+        bytes32 role,
+        TriState access
+    )
+        public
+    {
+        return setUserRole(subject, role, GLOBAL, access);
+    }
+
+    // Assign a role to a user globally
+    function setUserRole(
```

Comment:
Correct. You can always turn these two transactions into one using a helper wallet contract, but it will get a bit tricky with the rights to do so. I can easily add a batch operation:

```
    function setUserRole(
        address[] subjects,
        bytes32[] roles,
        IAccessControlled[] objects,
        TriState[] newValues
    )
```

Then you can do all sorts of complex operations atomically.